src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
```

Comment:
If we pass it to `handleAuthentication` and `matchesOp` as you suggested is slightly better, but it still has the problem that most `AuthenticationProvider` implementations do not use an instance of `ZooKeeperServer`.